regularly output some runtime stats

tbg opened the pull request:
> I0601 18:53:20.283745   29093 node.go:544] runtime stats: 42 goroutines, 26.27mb active, 706.16mb total, 14719 cgo/second

tbg committed:
regularly output some runtime stats

log2:I0601 18:53:20.283745   29093 node.go:544] runtime stats: 42 goroutines, 26.27mb active, 706.16mb total, 14719 cgo/second

petermattis commented on server/node.go:
```diff
@@ -517,6 +522,32 @@ func (n *Node) waitForScanCompletion() int64 {
+			case <-ticker.C:
+				runtime.ReadMemStats(&ms)
+				newCGOCall := runtime.NumCgoCall() - cgoCall
+				cgoRate := newCGOCall / (int64(runtimeStatsInterval) / int64(time.Second))
```
I'm not sure how accurate the tickers are. It might be better to keep track of the time between calls explicitly.

tschottdorf replied:
done.

petermattis commented on server/node.go:
```diff
@@ -517,6 +522,32 @@ func (n *Node) waitForScanCompletion() int64 {
+		defer ticker.Stop()
+		var cgoCall int64
+		ms := runtime.MemStats{}
+		mb := 10E-6
```
`const mb = 10e-6`

tschottdorf replied:
done.

petermattis commented on server/node.go:
```diff
@@ -517,6 +522,32 @@ func (n *Node) waitForScanCompletion() int64 {
 
+// startLogStats periodically outputs runtime statistics to the log files.
+func (n *Node) startRuntimeStats(stopper *util.Stopper) {
+	stopper.RunWorker(func() {
```
Add a `TODO(mtracy|bram)` to store these runtime stats into time series.

tschottdorf replied:
done.

petermattis commented on server/node.go:
```diff
@@ -517,6 +522,32 @@ func (n *Node) waitForScanCompletion() int64 {
 	return n.scanCount
 }
 
+// startLogStats periodically outputs runtime statistics to the log files.
```
`s/startLogStats/startRuntimeStats/g`

tschottdorf replied:
done.

spencerkimball commented on server/node.go:
```diff
@@ -46,6 +47,9 @@ const (
 	gossipInterval = 1 * time.Minute
+	// runtimeStatsInterval is the interval for logging runtime stats such
+	// as number of goroutines and memory allocated.
+	runtimeStatsInterval = 10 * time.Second
```
This seems a little busy. How about every 30s?

tschottdorf replied:
done.

tschottdorf replied:
Actually went back down to 15s. It's nice to have it a little more regular for now.

petermattis commented:
LGTM

spencerkimball commented on server/node.go:
```diff
@@ -517,6 +522,32 @@ func (n *Node) waitForScanCompletion() int64 {
 }
 
+// startLogStats periodically outputs runtime statistics to the log files.
+func (n *Node) startRuntimeStats(stopper *util.Stopper) {
```
Should these values be logged to stat counters?

tschottdorf replied:
done.

spencerkimball commented:
LGTM

tschottdorf commented:
added some more items from #1179 and addressed comments. PTALL

> I0601 21:10:45.767555   25134 node.go:556] runtime stats: 34 goroutines, 7.64mb/7.64mb/36.35mb active/heap/total, 79.13cgo/sec, 0.03/0.01 %(u/s)time

petermattis commented on server/node.go:
```diff
@@ -517,6 +524,42 @@ func (n *Node) waitForScanCompletion() int64 {
+					log.Errorf("Getrusage failed: %v", err)
+				}
+				uPerc := float64(ru.Utime.Nano()) / dur
+				sPerc := float64(ru.Stime.Nano()) / dur
```
Do you have to take the delta between `Getrusage()` calls?

tschottdorf replied:
I do. Fixing.

petermattis commented:
Still LGTM.

tamird commented on server/node.go:
```diff
@@ -517,6 +524,42 @@ func (n *Node) waitForScanCompletion() int64 {
+			case <-stopper.ShouldStop():
+				return
+			case <-ticker.C:
+				dur := float64(time.Now().Sub(tStart))
```
FYI http://godoc.org/time#Since

tschottdorf replied:
done.

tbg committed:
address feedback and add more data

tbg committed:
allow %% as literal percent sign

MBerhault commented on server/node.go:
```diff
@@ -46,6 +48,10 @@ const (
+	// runtimeStatsInterval is the interval for logging runtime stats such
+	// as number of goroutines and memory allocated.
+	runtimeStatsInterval = 15 * time.Second
+	mb                   = 10e-6
```
you probably want 1<<20 and float64(ms.XX) / mb

tschottdorf replied:
done.

MBerhault commented on server/node.go:
```diff
@@ -517,6 +524,46 @@ func (n *Node) waitForScanCompletion() int64 {
+				sPerc := float64(newStime-sTime) / dur
+				uTime = newUtime
+				sTime = newStime
+				log.Infof("runtime stats: %d goroutines, %.2fmb/%.2fmb/%.2fmb active/heap/total, %.2fcgo/sec, %.2f/%.2f %%(u/s)time",
```
change mb to MB for clarity

tschottdorf replied:
done

tbg committed:
additional feedback and add GC info

> runtime stats: 46 goroutines, 1.66MB
active, 65.35cgo/sec, 0.00/0.00 %(u/s)time, 0.00 %gc (0x)
